### DISCUSSION TITLE
Question: Quadtree thread safety

### DISCUSSION THREAD

**Discussion by thomasfrisko** [NetTopologySuite/NetTopologySuite#491] [Q&A]
Currently I'm using the spatial index tree structure Quadtree in my application where I'm frequently both updating and reading. To ensure thread safety I'm using a ReaderWriterLockSlim, however, as the load increases it creates a bottleneck due to frequent updates. I'm therefore wondering whether it is necessary to avoid concurrent read/writes or if the Quadtree structure already handles this.

**Comment by airbreather** [comment]
`NetTopologySuite.Index.Quadtree.Quadtree<T>` does not handle thread-safety for you automatically.  Ultimately, all successful inserts / deletes go through `List<T>.Add` / `List<T>.Remove`, respectively.

Depending on your data and usage patterns, perhaps you could try deferring updates by putting the details into a `ConcurrentBag<QuadtreeUpdate>`, and then periodically (on a timer) committing those updates to the tree in a batch?
